torchdata/stateful_dataloader/sampler.py

```diff
@@ -18,58 +19,162 @@ class _StatefulRandomSamplerIterator(Iterator[int], Stateful):
     _GENERATOR = "generator"
     _YIELDED = "yielded"
 
-    def __init__(self, sampler, parent_iterator: Iterator[int]):
+    def __init__(self, sampler):
         self.sampler = sampler
-        self.parent_iterator = parent_iterator
+        self.generator_state = self.sampler.generator.get_state()
         self.yielded = 0
         self.next_yielded = None
-        self.generator_state = sampler.generator.get_state()
+        self.n = len(sampler.data_source)
+        self.replacement = sampler.replacement
+        self.num_samples = sampler.num_samples
+        self.chunk_size = 32
+        self.chunk_index = 0
+        self.perm_index = 0
+        self.perm = None
 
-    def __next__(self) -> int:
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        if self.replacement:
+            num_full_chunks = self.num_samples // self.chunk_size
+            remainder = self.num_samples % self.chunk_size
+            if self.chunk_index < num_full_chunks:
+                if self.perm is None or not self.perm:
+                    self.perm = torch.randint(
+                        high=self.n,
+                        size=(self.chunk_size,),
+                        dtype=torch.int64,
+                        generator=self.sampler.generator,
+                    ).tolist()
+                    self.perm_index = 0
+                value = self.perm[self.perm_index]
+                self.perm_index += 1
+                if self.perm_index == self.chunk_size:
+                    self.chunk_index += 1
+                    self.perm = None
+                self.yielded += 1
+                return value
+            elif remainder > 0:
+                if self.perm is None or not self.perm:
+                    self.perm = torch.randint(
+                        high=self.n,
+                        size=(remainder,),
+                        dtype=torch.int64,
+                        generator=self.sampler.generator,
+                    ).tolist()
```

Comment:
Can we do implement with less repeated code and if/else branches? I think we need __next__ to be as lightweight as possible. 

I would suggest we generate the sequence once (ie at startup/load_state_dict or at the beginning of the function).

eg

self.seq = generate_sequence(replacement=True)
ret = self.seq[self.i]
self.i += 1
return ret

Reply:
Is the chunking logic present in random sampler?

Reply:
Yes, chunking is present in the torch.utils.data RandomSampler.